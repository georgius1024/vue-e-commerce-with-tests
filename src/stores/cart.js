import { defineStore } from 'pinia'

export const useCartStore = defineStore('cart', {
  state: () => {
    return { products: {} }
  },
  getters: {
    cart(state) {
      return Object.values(state.products)
    },
    size(state) {
      return Object.keys(state.products).length
    },
    inCart() {
      return (id) => Boolean(this.products[id])
    }
  },
  actions: {
    add(product) {
      console.log(JSON.stringify(product), JSON.stringify(this.products))
      this.products[product.id] = product
    },
    remove(product) {
      delete this.products[product.id]
    }
  },
})