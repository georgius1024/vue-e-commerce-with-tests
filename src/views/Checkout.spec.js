import { describe, it, expect } from "vitest";
import { mount } from "@vue/test-utils";
import { setActivePinia, createPinia } from "pinia";
import { useCartStore } from "@/stores/cart";
import component from "@/views/Checkout.vue";

describe("Checkout page", () => {
  const products = global.productsMock
  const buildWrapper = () => {
    return mount(component);
  };
  beforeEach(() => {
    setActivePinia(createPinia());
    const store = useCartStore();
    products.forEach(product => store.add(product))
  });
  it("mounts without error",  () => {
    const wrapper = buildWrapper();
    expect(wrapper.exists).toBeTruthy();
    expect(wrapper.html()).toMatchSnapshot();
  });

  it("click on remove control calls store action", async () => {
    const store = useCartStore();
    const remove = vi.spyOn(store, 'remove')
    const wrapper = buildWrapper();
    expect(wrapper.exists).toBeTruthy();
    await wrapper.find('[data-testid="remove-product-101"').trigger("click");
    expect(remove).toHaveBeenCalled()
  });

  it("click on remove control removes product", async () => {
    const wrapper = buildWrapper();
    expect(wrapper.findAll('li')).toHaveLength(3)
    await wrapper.find('[data-testid="remove-product-101"').trigger("click");
    expect(wrapper.findAll('li')).toHaveLength(2)
  });

  it("click on checkout calls global method", async () => {
    const wrapper = buildWrapper();
    const alert = vi.fn()
    global.alert = alert    
    await wrapper.find('.btn-primary').trigger("click");
    expect(alert).toHaveBeenCalled()
  });

});
