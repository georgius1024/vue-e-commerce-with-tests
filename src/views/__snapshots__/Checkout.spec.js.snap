// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Checkout page > mounts without error 1`] = `
"<div class=\\"container\\">
  <h1>Products in your cart</h1>
  <div class=\\"card w-50\\">
    <ul class=\\"list-group list-group-flush\\">
      <li class=\\"list-group-item text-capitalize\\"><img class=\\"me-3\\" width=\\"48\\"> product-1 <span class=\\"float-end text-danger pa-3\\" style=\\"cursor: pointer;\\" data-testid=\\"remove-product-101\\"> × </span></li>
      <li class=\\"list-group-item text-capitalize\\"><img class=\\"me-3\\" width=\\"48\\"> product-2 <span class=\\"float-end text-danger pa-3\\" style=\\"cursor: pointer;\\" data-testid=\\"remove-product-102\\"> × </span></li>
      <li class=\\"list-group-item text-capitalize\\"><img class=\\"me-3\\" width=\\"48\\"> product-3 <span class=\\"float-end text-danger pa-3\\" style=\\"cursor: pointer;\\" data-testid=\\"remove-product-103\\"> × </span></li>
    </ul>
  </div><button class=\\"w-50 mt-3 btn btn-primary\\"> Checkout </button>
</div>"
`;
